File: src/Java/algorithms/mathandprob/RomanToInteger.java
package algorithms.mathandprob; 

import java.util.*; 

public class RomanToInteger {
    /* Function: Given a roman number, convert it into integer. 
     * Solution: Each time add the current value. Notice only 4, 9
     *     does not obey the role, so use minus double: IV=1+5-2.
     */ 
    
    public int romanToInt(String s) {
        if (s == null || s.length() == 0) {
            return 0; 
        }
        
        int sum = 0, pre = 0; 
        for (int i = 0; i < s.length(); ++i) {
            int cur = romanList(s.charAt(i)); 
            sum += cur; 
            if (cur > pre) {
                sum -= 2 * pre; 
            }
            pre = cur; 
        }
        
        return sum; 
    }
    
    private int romanList(char c) {
        switch (c) {
            case 'I': return 1; 
            case 'V': return 5; 
            case 'X': return 10; 
            case 'L': return 50; 
            case 'C': return 100; 
            case 'D': return 500; 
            case 'M': return 1000; 
            default: return 0; 
        }
    }
}

File: wwwroot/js/roman-to-integer.js
/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function(s) {
    let romanVal = {
        'I': 1,
        'V': 5,
        'X': 10,
        'L': 50,
        'C': 100,
        'D': 500,
        'M': 1000
    };

    let sum = 0;
    let prev = 0;
    for (let i = 0; i < s.length; i++) {
        let cur = romanVal[s[i]];
        if (cur > prev) {
            sum -= 2 * prev;
        }

        sum += cur;
        prev = cur;
    }

    return sum;
};

document.getElementById("roman-to-integer").innerHTML = romanToInt("IX");